adapters/eplanning/eplanning.go

```diff
@@ -0,0 +1,152 @@
+package eplanning
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"fmt"
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/openrtb_ext"
+
+	"strconv"
+)
+
+type EPlanningAdapter struct {
+	http *adapters.HTTPAdapter
+	URI  string
+}
+
+func (adapter *EPlanningAdapter) MakeRequests(request *openrtb.BidRequest) ([]*adapters.RequestData, []error) {
+	errors := make([]error, 0, len(request.Imp))
+	totalImps := len(request.Imp)
+	sourceMapper := make(map[string][]int)
+
+	for i := 0; i < totalImps; i++ {
+		source, err := verifyImp(&request.Imp[i])
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		if _, ok := sourceMapper[source]; !ok {
+			sourceMapper[source] = make([]int, 0, totalImps-i)
+		}
+
+		sourceMapper[source] = append(sourceMapper[source], i)
+	}
+
+	totalRequests := len(sourceMapper)
+
+	if totalRequests == 0 {
+		return nil, errors
+	}
+
+	requests := make([]*adapters.RequestData, 0, totalRequests)
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json")
+	headers.Add("Accept", "application/json")
+	if request.Device != nil {
+		addHeaderIfNonEmpty(headers, "User-Agent", request.Device.UA)
+		addHeaderIfNonEmpty(headers, "X-Forwarded-For", request.Device.IP)
+		addHeaderIfNonEmpty(headers, "Accept-Language", request.Device.Language)
+		addHeaderIfNonEmpty(headers, "DNT", strconv.Itoa(int(request.Device.DNT)))
+	}
+
+	imps := request.Imp
+
+	for source, impIds := range sourceMapper {
+		request.Imp = request.Imp[:0]
+
+		for i := 0; i < len(impIds); i++ {
+			request.Imp = append(request.Imp, imps[impIds[i]])
```

Comment:
I _think_  there's a bug here, although I'm not certain of it. The `imps := request.Imp` line above only makes a copy of the _slice_--which is basically just a pointer.

So then `imps` and `request.Imp` are pointing to the same data. When you call:

```
request.Imp = request.Imp[:0]
request.Imp = append(request.Imp, {something})
```

then you're actually overriding the first element in memory. If you call `append` twice, you'd override the first and second.

As you loop through the elements, the data `imps` points to will be changing... which may then screw up your `imps[impIds[i]]` call, because you won't be fetching the `Imp` you think you are.

It would be very hard to pick this up by testing, because the `range map` iteration order isn't guaranteed, and you'd need the _second_ element it returns to use an `imp` from the _first_ index.

Does that make sense?

Reply:
Here's a playground snippet which illustrates the issue: https://play.golang.org/p/MYF9LIOExiv